azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```diff
@@ -5,150 +5,166 @@
 # --------------------------------------------------------------------------
 """ This module contains the manager for handler methods used by the aio clients"""
 
-from azure.iot.device.common import asyncio_compat
-from azure.iot.device.common.chainable_exception import ChainableException
-from concurrent.futures import ThreadPoolExecutor
 import asyncio
 import logging
 import inspect
+import threading
+import concurrent.futures
+from azure.iot.device.common import asyncio_compat, handle_exceptions
+from azure.iot.device.iothub.sync_handler_manager import (
+    AbstractHandlerManager,
+    HandlerManagerException,
+    HandlerRunnerKillerSentinel,
+)
 
 logger = logging.getLogger(__name__)
 
-MESSAGE = "_on_message_received"
-METHOD = "_on_method_request_received"
-TWIN_DP_PATCH = "_on_twin_desired_properties_patch_received"
-# TODO: add more for "event"
+# This logic could potentially be encapsulated in another module...
+# Say... some kind of... loop_manager.py?
+RUNNER_LOOP = asyncio.new_event_loop()
+RUNNER_THREAD = threading.Thread(target=RUNNER_LOOP.run_forever)
+RUNNER_THREAD.daemon = True
+RUNNER_THREAD.start()
 
 
-class AsyncHandlerManagerException(ChainableException):
-    pass
+class AsyncHandlerManager(AbstractHandlerManager):
+    """Handler manager for use with asynchronous clients"""
 
-
-class AsyncHandlerManager(object):
-    def __init__(self, inbox_manager):
-        self._inbox_manager = inbox_manager
-
-        self._handler_tasks = {
-            # Inbox handler tasks
-            MESSAGE: None,
-            METHOD: None,
-            TWIN_DP_PATCH: None,
-            # Other handler tasks
-            # TODO: add
-        }
-
-        # Inbox handlers
-        self._on_message_received = None
-        self._on_method_request_received = None
-        self._on_twin_desired_properties_patch_received = None
-
-        # Other handlers
-        # TODO: add
-
-    async def _run_inbox_handler(self, inbox, handler_name):
+    async def _inbox_handler_runner(self, inbox, handler_name):
         """Run infinite loop that waits for an inbox to receive an object from it, then calls
         the handler with that object
         """
+        logger.debug("HANDLER RUNNER ({}): Starting runner".format(handler_name))
+
+        # Define a callback that can handle errors in the ThreadPoolExecutor
+        def _handler_callback(future):
+            try:
+                e = future.exception()
+            except concurrent.futures.CancelledError as raised_e:
+                new_err = HandlerManagerException(
+                    message="HANDLER ({}): Invocation unexpectedly ended with cancellation".format(
+                        handler_name
+                    ),
+                    cause=raised_e,
+                )
+                handle_exceptions.handle_background_exception(new_err)
```

Comment:
>handle_exceptions.handle_background_exception(new_err) [](start = 16, length = 54)

If, by some chance, future.exception raises some other error, (which it _shouldn't_  AFAIK), that error will be swallowed.  Suggest adding a catch-all here instead of just catching CancelledError.